project_role/models/project_assignment.py

```diff
@@ -18,50 +21,111 @@ class ProjectAssignment(models.Model):
         (
             'project_role_user_uniq',
-            'unique (project_id, role_id, user_id)',
+            'UNIQUE (project_id, role_id, user_id)',
             'User may be assigned per role only once within a project!'
         ),
+        (
+            'company_role_user_uniq',
+            (
+                'EXCLUDE ('
+                '    company_id WITH =, role_id WITH =, user_id WITH ='
+                ') WHERE ('
+                '    project_id IS NULL'
+                ')'
+            ),
+            'User may be assigned per role only once within a company!'
+        ),
+        (
+            'nocompany_role_user_uniq',
+            (
+                'EXCLUDE (role_id WITH =, user_id WITH =) WHERE ('
+                '    project_id IS NULL AND company_id IS NULL'
+                ')'
+            ),
+            'User may be assigned per role only once!'
+        ),
     ]
 
-    @api.depends('project_id.name', 'role_id.name', 'user_id.name')
+    @api.depends(
+        'company_id.name',
+        'project_id.name',
+        'role_id.name',
+        'user_id.name',
+    )
     def _compute_name(self):
         for assignment in self:
-            assignment.name = _('%s as %s on %s') % (
-                assignment.user_id.name,
-                assignment.role_id.name,
-                assignment.project_id.name,
-            )
+            if assignment.project_id:
+                assignment.name = _('%s as %s on %s') % (
+                    assignment.user_id.name,
+                    assignment.role_id.name,
+                    assignment.project_id.name,
+                )
+            elif assignment.company_id:
+                assignment.name = _('%s as %s in %s') % (
+                    assignment.user_id.name,
+                    assignment.role_id.name,
+                    assignment.company_id.name,
+                )
+            else:
+                assignment.name = _('%s as %s') % (
+                    assignment.user_id.name,
+                    assignment.role_id.name,
+                )
+
+    @api.multi
+    def _get_conflicting_domain(self):
+        self.ensure_one()
+        return [
+            ('id', '!=', self.id),
+            ('role_id', '=', self.role_id.id),
+            ('user_id', '=', self.user_id.id),
+        ] + (
+            [('company_id', 'in', [False, self.company_id.id])]
+            if self.company_id else []
+        ) + (
+            [('project_id', 'in', [False, self.project_id.id])]
+            if self.project_id else []
+        )
 
     @api.multi
-    @api.constrains('role_id', 'user_id')
-    def _check_assignable(self):
+    @api.constrains('company_id', 'project_id', 'role_id', 'user_id')
+    def _check(self):
         for assignment in self:
+            conflicting_assignment = self.search(
+                assignment._get_conflicting_domain(),
+                limit=1,
+            )
+            if conflicting_assignment:
```

Comment:
Isn't this redundant considering the constraint. Or do I misunderstand it :smile: maybe

Reply:
It's only a bit redundant, in SQL constraint it's hard (if even possible) to express logic like "if there's global assignment of user U to role R, don't allow project-specific assignment". So I could've removed SQL constraints but I think it's good to keep things consistent as much as possible on database level as well :)